src/core/meshcop/border_agent.cpp

```diff
@@ -308,23 +309,38 @@ Error BorderAgent::SetId(const Id &aId)
 
 void BorderAgent::HandleNotifierEvents(Events aEvents)
 {
-    VerifyOrExit(aEvents.ContainsAny(kEventThreadRoleChanged | kEventCommissionerStateChanged));
-
+    if (aEvents.ContainsAny(kEventThreadRoleChanged | kEventCommissionerStateChanged))
+    {
 #if OPENTHREAD_CONFIG_COMMISSIONER_ENABLE && OPENTHREAD_FTD
-    VerifyOrExit(Get<Commissioner>().IsDisabled());
+        VerifyOrExit(Get<Commissioner>().IsDisabled());
 #endif
 
-    if (Get<Mle::MleRouter>().IsAttached())
-    {
-        Start();
+        if (Get<Mle::MleRouter>().IsAttached())
+        {
+            Start();
+        }
+        else
+        {
+            Stop();
+        }
     }
-    else
+
+    if (aEvents.ContainsAny(kEventPskcChanged))
     {
-        Stop();
+        if (mState == kStateConnected || mState == kStateAccepted)
+        {
+            mPskcChangedDuringConnectedOrAccepted = true;
+        }
+        else if (mState == kStateStarted)
+        {
+            RestartToApplyNewPskc();
```

Comment:
This is similar to restarting agent after e-PSK? Maybe we can combine the methods into one?
(can consider adding a `Reason` enum to indicate why we are restarting if we need to take any different actions)?

Reply:
did have considered it, however chosen this after considering the difference when submitting this PR.
It does make sense to combine similar code, let me update.